apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,267 @@
+
+  // Find the <script> tag containing the location metadata arrays
+  // Example: <script>let locationIds = ['245', '677', ...]; let floors = ['-GSC-2', '-SL-3', ...];</script>
+  const scriptText = $("script")
+    .toArray()
+    .map((el) => $(el).html() ?? "")
+    .find((txt) => /let\s+locationIds\s*=\s*\[/s.test(txt));
+
+  if (!scriptText) throw new Error("Could not find <script> containing locationIds");
+
+  // Extract the content of the locationIds array from the script
+  // Regex captures everything between [ and ], e.g., "'245', '677', '205'"
+  const locMatch = scriptText.match(/let\s+locationIds\s*=\s*\[([\s\S]*?)]/);
+  // Extract the content of the floors array from the script
+  // Regex captures everything between [ and ], e.g., "'-GSC-2', '-SL-3', '1'"
+  const floorMatch = scriptText.match(/let\s+floors\s*=\s*\[([\s\S]*?)]/);
+  if (!locMatch || !floorMatch) throw new Error("Unable to capture locationIds or floors array");
+
+  // Extract parallel arrays:
+  // - locationIds: unique Occuspace IDs for each location (i.e. ['245', '677', '205'])
+  // - floorCodes: corresponding internal DOM codes for each location's container (i.e. ['-GSC-2', '-SL-3', '1'])
+  const locationIds = parseArray(locMatch[1]);
+  const floorCodes = parseArray(floorMatch[1]);
+
+  // These arrays must have matching lengths because each location ID corresponds to exactly one floor code
+  // If lengths differ, the HTML structure changed or the regex extraction failed, and we should fail loudly
+  if (locationIds.length !== floorCodes.length)
+    throw new Error("locationIds and floorCodes arrays are different lengths");
+
+  const locationMeta: Record<string, LocationMeta> = {};
+
+  locationIds.forEach((id, idx) => {
+    const floorCode = floorCodes[idx];
+    // Locate the DOM container for this location using the floor code
+    const locationSelection = $(`#leftSide${floorCode}`);
+
+    // Extract library name from card title (e.g., "Langson Library", "Science Library")
+    const libraryLabel = locationSelection.find("h2.card-title").first().text().trim();
+    // Extract sub-location label if present (e.g., floor/section name)
+    const subLocationLabel = locationSelection
+      .find("span.subLocation")
+      .first()
+      .text()
+      .trim()
+      .replace(/\s+/g, " ");
+    const fullLocationLabel = subLocationLabel
+      ? `${libraryLabel} - ${subLocationLabel}`
+      : libraryLabel || "Unknown";
+
+    locationMeta[id] = {
+      id,
+      floorCode: floorCode,
+      libraryName: codeToLibrary(floorCode),
+      locationLabel: fullLocationLabel,
+    };
+  });
+
+  return locationMeta;
+}
+
+async function fetchLocation(id: string): Promise<RawRespOK["data"] | null> {
+  const url = `https://www.lib.uci.edu/sites/all/scripts/occuspace.php?id=${id}`;
+  try {
+    // Double-encoded JSON: double parse required
+    const parsedResponse = await fetch(url)
+      .then((r) => r.text())
+      .then(JSON.parse)
```

Comment:
Why collect the entire string just to `JSON.parse` it? `Response.json` exists so that it can work on the TCP stream instead of temporarily allocating the entire response body.
```suggestion
      .then((r) => r.json())
```